test/metabase/lib/fe_util_test.cljc

```diff
@@ -125,6 +135,110 @@
+                    {:operator :case
+                     :options {}
+                     :args [boolean-field int-field]}
+
+                    (lib/case [[boolean-field int-field]] nil)
+                    {:operator :case
+                     :options {}
+                     :args [boolean-field int-field]}
+
+                    (lib/case [[boolean-field int-field]] string-field)
+                    {:operator :case
+                     :options {}
+                     :args [boolean-field int-field string-field]}
+
+                    (lib/case [[boolean-field int-field] [boolean-field string-field]])
+                    {:operator :case
+                     :options {}
+                     :args [boolean-field int-field boolean-field string-field]}
+
+                    (lib/case [[boolean-field int-field] [boolean-field string-field]] nil)
+                    {:operator :case
+                     :options {}
+                     :args [boolean-field int-field boolean-field string-field]}
+
+                    (lib/case [[boolean-field int-field] [boolean-field string-field]] dt-field)
+                    {:operator :case
+                     :options {}
+                     :args [boolean-field int-field boolean-field string-field dt-field]}}]
+    (testing "case pairs should be flattened in expression parts"
+      (doseq [[clause parts] test-cases]
+        (let [{:keys [operator options args]} parts
+              res (lib.fe-util/expression-parts query -1 clause)]
+          (is (=? operator (:operator res)))
+          (is (=? options  (:options res)))
+          (is (=? (map :id args) (map :id (:args res)))))))
+
+    (testing "case pairs should be unflattened in expression clause"
+      (doseq [[expected-expression parts] test-cases]
+        (let [{:keys [operator options args]} parts
+              actual-expression (lib.fe-util/expression-clause operator args options)
+              [expected-op _ expected-args] expected-expression
+              [actual-op _ actual-args] actual-expression]
+          (is (=? expected-op actual-op))
+          (is (=? (map :id expected-args) (map :id actual-args))))))
+
+    (testing "case/if should round-trip through expression-parts and expression-clause"
+      (doseq [[clause] test-cases]
+        (let [parts                     (lib.fe-util/expression-parts query clause)
+              round-tripped-expression  (lib.fe-util/expression-clause
+                                         (:operator parts)
+                                         (:args parts)
+                                         nil)
+              round-tripped-parts       (lib.fe-util/expression-parts query round-tripped-expression)]
+
+          (is (=? (:operator parts) (:operator round-tripped-parts)))
+          (is (=? (map :id (:args parts)) (map :id (:args round-tripped-parts)))))))
+
+    (testing "deeply nested case/if should round-trip through expression-parts and expression-clause"
```

Comment:
This last `testing` doesn't seem to use the shared stuff, aside from the trivial bits like the `boolean-field` etc.
May as well split it into a separate `deftest` that easier to read.